### PR TITLE
feat(gpu): implement signed_overflowing_add

### PR DESCRIPTION
**Pull request by bbarbakadze**
### PR content/description

Implements signed overflowing add in cuda backend.

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: backends/tfhe-cuda-backend/cuda/src/integer/addition.cuh [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  // phase 3
+  auto input_carry = &input_carries[(num_blocks - 1) * big_lwe_size];
+
+  host_resolve_signed_overflow(
+      streams, gpu_indexes, gpu_count, overflowed, last_block_inner_propagation,
+      input_carry, output_carry, mem_ptr->resolve_overflow_mem, bsks, ksks);
+
+  cuda_memcpy_async_gpu_to_gpu(lhs, result, num_blocks * big_lwe_size_bytes,
+                               streams[0], 0);
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/integer/addition.cuh:141]: You should replace:
```
streams[0], 0);
```
by: 
```
streams[0], gpu_indexes[0]);
```
With a single GU the user has the possibility to pass a GPU index different from 0.

File: tfhe/src/integer/gpu/server_key/radix/add.rs [hunk 2]
```diff
@@ -491,4 +495,202 @@ impl CudaServerKey {
 
         (ct_res, ct_overflowed)
     }
+
+    pub fn signed_overflowing_add(
```
Comment (agnesLeroy) [inline tfhe/src/integer/gpu/server_key/radix/add.rs:499]: Here could you add a doc test to give an example of use please?

**Review by agnesLeroy — COMMENTED** [review]
Hey @bbarbakadze! Thanks a lot for this PR, here comes my review :slightly_smiling_face:

File: tfhe/benches/integer/signed_bench.rs [hunk 3]
```diff
@@ -1938,6 +1943,11 @@ mod cuda {
         display_name: max
     );
 
+    define_cuda_server_key_bench_clean_input_signed_fn!(
+        method_name: signed_overflowing_add,
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/benches/integer/signed_bench.rs:1947]: Not really, it's in `display_name:` that the prefix `signed_` **must not** appear.

File: tfhe/benches/integer/signed_bench.rs [hunk 4]
```diff
@@ -1938,6 +1943,11 @@ mod cuda {
...
+        display_name: signed_overflowing_add
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/benches/integer/signed_bench.rs:1948]: Here you're right.

File: tfhe/benches/integer/signed_bench.rs [hunk 5]
```diff
@@ -2068,6 +2078,7 @@ mod cuda {
         cuda_unchecked_le,
         cuda_unchecked_min,
         cuda_unchecked_max,
+        cuda_unchecked_signed_overflowing_add,
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/benches/integer/signed_bench.rs:2081]: We're already in `signed_bench.rs` file so no need to be redundant.

File: backends/tfhe-cuda-backend/cuda/include/integer.h [hunk 6]
```diff
@@ -35,8 +35,11 @@ enum COMPARISON_TYPE {
   MAX = 6,
   MIN = 7,
 };
+
 enum CMP_ORDERING { IS_INFERIOR = 0, IS_EQUAL = 1, IS_SUPERIOR = 2 };
 
+enum SIGNED_OPERATION { ADDITION = 1, SUBTRACTION = -1 };
```
[the remark this answers is not in the document]
Comment (bbarbakadze) [reply backends/tfhe-cuda-backend/cuda/include/integer.h:41]: it is subtraction or addition, not related to overflow and on rust side it has the same name so maybe we keep it

[the remark this answers is not in the document]
Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/include/integer.h:41]: Ok no problem :slightly_smiling_face:

File: backends/tfhe-cuda-backend/cuda/src/integer/addition.cuh [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/integer/addition.cuh:141]: Here the change is missing it seems

File: backends/tfhe-cuda-backend/cuda/include/integer.h [hunk 8]
```diff
@@ -2884,6 +2905,229 @@ template <typename Torus> struct int_div_rem_memory {
...
+    };
+
+    last_block_inner_propagation_lut = new int_radix_lut<Torus>(
+        streams, gpu_indexes, gpu_count, params, 1, 1, allocate_gpu_memory);
+
+    generate_device_accumulator_bivariate<Torus>(
+        streams[0], gpu_indexes[0],
+        last_block_inner_propagation_lut->get_lut(gpu_indexes[0], 0),
+        params.glwe_dimension, params.polynomial_size, message_modulus,
+        params.carry_modulus, f_last_block_inner_propagation_lut);
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/include/integer.h:2970]: Here you need to broadcast the lut for multi-gpu support

**Review by agnesLeroy — APPROVED** [review]

**Pull request merged**